Added tpa and tpaccept

JerryHigarr opened the pull request:
I have been coding a large number of commands for my vanilla server using this project, so I wanted to show my appreciation by offering a couple of them :)

JerryHigarr committed:
Added tpa and tpaccept

I have been coding a large number of commands for my vanilla server using this project, so I wanted to show my appreciation by offering a couple of them :)

Gjum commented on plugins/base_commands.py:
```diff
@@ -23,11 +27,34 @@ class BaseCommandsPlugin(PluginBase):
+                logger.debug(json.dumps(tpareqs))
+                del tpareqs[towho] #clear the table.
+            else:
+                del tpareqs[towho] #clear the table.
```
Uh oh, code duplication. :)

nickelpro commented on plugins/base_commands.py:
```diff
@@ -8,8 +8,12 @@
 from spockbot.plugins.base import PluginBase
 from spockbot.vector import Vector3
+import json
+tpareqs = {}
```
Why is this a global. Like, I'm all for throwing coding practice to the wind, just ask @txomon or @gamingrobot. But for next time, random module scoped globals aren't a super great thing for me to see in a PR.

txomon replied:
if tpareqs is just used in the module, it should be a class level object. as long as you just modify it, all instances from the class will have access to it. If overwritten, just the object will loose the reference to it. Be carefull with the lifetime of the object, it will never be destroyed!

nickelpro replied:
Nah it should be an instanced variable set up in init. No need to share between objects. This code will explode hard if you ever tore down the plugin and then recovered (same for class scope). 

That's maybe a bit of an esoteric example but correct scoping is a super important programming instinct to learn. Code smells and all that

JerryHigarr replied:
Fair enough, thanks for the tips. Hopefully this can be useful to somebody to some end :)

JerryHigarr replied:
Fixing it now, one sec.

Bc1151 commented:
Fair enough, thanks for the tips. Hopefully this can be useful to somebody to some end :)

Bc1151 commented:
Fixing it now, one sec.

JerryHigarr committed:
cleaned it up.

hopefully that's a tad more acceptable.

JerryHigarr commented:
Hopefully that's a tad more acceptable :)

txomon commented on plugins/base_commands.py:
```diff
@@ -23,11 +27,28 @@ class BaseCommandsPlugin(PluginBase):
     def __init__(self, ploader, settings):
         super(BaseCommandsPlugin, self).__init__(ploader, settings)
-
+    def handle_tpa(self, event, data):
```
Add one more space before (use flake8 for checking all the file)

JerryHigarr replied:
Oh, I'll get that now

txomon commented:
Apart from the formatting, it looks good to me, but I haven't tested it.

JerryHigarr committed:
Clean up.

Again, a big thanks to @Gjum who has been a big help in my first steps with python :)

JerryHigarr commented:
I just tested it, and I fixed some things. A big thanks to @Gjum who has been very understanding.

JerryHigarr committed:
flake8 fix.

thanks @xomon for the pointer :)

JerryHigarr commented:
added a flake8 fix thanks to @txomon :)

txomon commented:
I meant that you should run flake8 for the formatting stuff:

```
$ flake8 plugins/base_commands.py 
plugins/base_commands.py:41:80: E501 line too long (156 > 79 characters)
plugins/base_commands.py:43:80: E501 line too long (126 > 79 characters)
plugins/base_commands.py:44:5: E301 expected 1 blank line, found 0
plugins/base_commands.py:48:80: E501 line too long (100 > 79 characters)
plugins/base_commands.py:50:38: E261 at least two spaces before inline comment
plugins/base_commands.py:50:39: E262 inline comment should start with '# '
plugins/base_commands.py:52:80: E501 line too long (124 > 79 characters)
plugins/base_commands.py:53:5: E301 expected 1 blank line, found 0
```

JerryHigarr committed:
actual flak8 fixes

thanks @txomon 
a special thanks to the entire SpockBot team, you guys are super nice :)

JerryHigarr commented:
There we go :) - thank you guys so much for being understanding with a beginner :)
(for what it matters, flake8 still spits this at me: "unknown option 'application-import-names' ignored
")

Gjum commented:
Is `''.join(data['name'])` really needed? Same with the other join. What really happens there is you split the string in chars for iteration, and then join the chars to get the original string back.

JerryHigarr committed:
removed redundant joins

JerryHigarr commented:
Fixed, thanks.

JerryHigarr commented:
anything else you need me to fix on this? or is it ready to merge?

txomon commented:
I can't accept it (I don't control this part), but here are some more comments on the patch, which are just observations and recommendations, is up to @gamingrobot to then merge it.

In handle_tpa you create args from data, but you just use it [0] from it. Apart from not being clear what 'args' is meant to represent, would it maybe more clear if you created something like:

```
from_who = data['args'][0]
```

And later you just have to use it from_who

One more recommendation I usually do is to keep in the first indentation the normal flow, and leave other cases in secondary levels, for example (did this without formatting):

```
    def handle_tpa(self, event, data):
        if not data['args']:
            self.net.push_packet('PLAY>Chat Message',
                                 {'message': '/tell ' + data['name'] +
                                  ' Usage: (!)tpa [name]'})
            return
        from_who = data['args'][0]
        self.tpa_reqs[from_who] = data['name']
        self.net.push_packet('PLAY>Chat Message',
                             {'message': '/tell %s would like to tpa to you, type (!)tpaccept or (!)tpdeny' % (from_who)})
```

That way you have a more clear path. Last but not least, I have seen the request, the accept, but I haven't seen the deny =)
